pkg/pool-stable/test/ComposableStablePool.test.ts

```diff
@@ -605,6 +607,124 @@ describe('ComposableStablePool', () => {
+                .div(fp(1).sub(expectedProtocolOwnershipPercentage));
+            });
+
+            it('returns the total supply after protocol fees are paid', async () => {
+              const virtualSupply = await pool.getVirtualSupply();
+              const expectedVirtualSupplyAfterProtocolFees = virtualSupply.add(expectedBptAmount);
+              const { preJoinExitSupply } = await pool.instance.callStatic.beforeJoinExit(registeredBalancesWithFees);
+              expect(preJoinExitSupply).to.be.almostEqual(expectedVirtualSupplyAfterProtocolFees, FEE_RELATIVE_ERROR);
+            });
+          });
+
+          context('when no protocol fees are due', () => {
+            it('returns the total supply', async () => {
+              const virtualSupply = await pool.getVirtualSupply();
+              const { preJoinExitSupply } = await pool.instance.callStatic.beforeJoinExit(registeredBalances);
+              expect(preJoinExitSupply).to.be.eq(virtualSupply);
+            });
+          });
+        });
+
+        it('returns the balances array having dropped the BPT balance', async () => {
+          const expectedBalances = registeredBalances.filter((_, i) => i !== bptIndex);
+          const { balances } = await pool.instance.callStatic.beforeJoinExit(registeredBalances);
+          expect(balances).to.be.deep.eq(expectedBalances);
+        });
+
+        it('returns the current amp factor', async () => {
+          const { value: expectedAmp } = await pool.getAmplificationParameter();
+          const { currentAmp } = await pool.instance.callStatic.beforeJoinExit(registeredBalances);
+          expect(currentAmp).to.be.deep.eq(expectedAmp);
+        });
+
+        it('returns the pre-join invariant', async () => {
+          const { value, precision } = await pool.getAmplificationParameter();
+          const expectedInvariant = calculateInvariant(
+            registeredBalances.filter((_, i) => i !== bptIndex),
+            value.div(precision)
+          );
+
+          const { preJoinExitInvariant } = await pool.instance.callStatic.beforeJoinExit(registeredBalances);
+
+          // We allow the invariant to experience a rounding error during calculation, however we want to make sure that
+          // we don't accept the invariant calculated using the old amplification factor so we use this strict check.
+          const error = preJoinExitInvariant.sub(expectedInvariant).abs();
+          expect(error).to.be.lte(1);
+        });
+      }
+
+      context('when the amplification factor is unchanged from the last join/exit', () => {
+        itPaysProtocolFeesAndReturnsNecessaryData();
+      });
+
+      context('when the amplification factor has changed from the last join/exit', () => {
+        context('when the amplification factor update is ongoing', () => {
+          sharedBeforeEach('perform an amp update', async () => {
+            await pool.startAmpChange(AMPLIFICATION_PARAMETER.mul(2));
```

Comment:
Let's pass the `duration` argument here - otherwise we need to remember that we default to two days.